src/Kernel/Process.class.st

```diff
@@ -213,63 +213,46 @@ Process >> debugWithTitle: title [
-	ctxt := suspendedContext.
-	ctxt isUnwindContext ifFalse: [ctxt := ctxt findNextUnwindContextUpTo: nil].
-	[ctxt isNil] whileFalse: [
-		ctxt unwindComplete ifNil: [
-			ctxt unwindComplete: true.
-			unwindBlock := ctxt unwindBlock.
-			"Create a context for the unwind block and execute it on the unwind block's stack. 
-			Note: using #value instead of #runUntilErrorOrReturnFrom: would lead to executing 
-			the unwind on the wrong stack preventing the correct execution of non-local returns."
-			suspendedContext := unwindBlock asContextWithSender: ctxt.
-			suspendedContext runUntilErrorOrReturnFrom: suspendedContext].
-		ctxt := ctxt findNextUnwindContextUpTo: nil].
-
-	"Mark the context as terminated"
-	suspendedContext terminateTo: nil.
-	suspendedContext setSender: nil receiver: self method: (Process>>#endProcess) arguments: {}
+	It assumes that self is not the active process and the termination is requested from another process"	
+		
+	| ctxt oldList outerMost top newTop |
+
+	[ ] ensure: [ "Execute termination as an ensure block to ensure it completes even if terminated by another process before the termination is finished; see testTerminateInTerminate."
+		self isTerminating ifTrue: [ ProcessAlreadyTerminating signal. ^self ].
+		terminating := true.
+		"Always suspend the process first so it doesn't accidentally get woken up"
+		oldList := self suspend.
+		suspendedContext ifNil: [^self].
+		"Figure out if we are terminating the process while waiting inside special object like in Semaphore>>critical:. So if waiting object is interesting on this situation we will ask it to handle it. In case of Semaphore>>critical, Semaphore will pop the suspendedContext so that we leave the ensure: block inside Semaphore>>critical: without signaling the semaphore.
+		This methods allow to not be restricted only on Semaphore case."
+		suspendedContext := oldList handleProcessTerminationOfWaitingContext: suspendedContext.
+
+		top := suspendedContext.
+		suspendedContext := nil.  "Disable terminating process while running its stack in active process below.
+		If terminating a process halfways through an unwind, try to complete that unwind block first; if there are multiple such nested unwind blocks, try to complete the outer-most one; nested unwind blocks will be completed in the process; see testTerminationDuringUnwind, testNestedUnwind.
+		Note: Halfway-through blocks have already set the complete variable (unwindComplete) in their defining #ensure:/#ifCurtailed contexts from nil to true; we'll search for the bottom-most one.
+		Note: #findNextUnwindContextUpTo: starts searching from the receiver's sender but the receiver 	itself may be an unwind context (see testTerminateEnsureAsTopContext)."
+		ctxt := top.
+		ctxt isUnwindContext ifFalse: [ctxt := ctxt findNextUnwindContextUpTo: nil].
+		[ctxt isNil] whileFalse: [
+			ctxt unwindComplete ifNotNil: [
+				outerMost := ctxt].
+			ctxt := ctxt findNextUnwindContextUpTo: nil].
```

Comment:
Let's introduce Context>>findMostOuterUnwindContext

Reply:
I haven't made my mind about this yet. There might be some more changes here in the future and splitting the method into smaller ones makes it, for me at least, less readable. I need to see the whole code in this case :) (I don't do this generally but in this particular case with a tricky code like #teminate I prefer less fragmented code). When the code "settles" I'm definitely not against fragmenting. I just know how many times I reshuffled this code which is why I'm presenting it as an unfragmented method.